freqtrade/worker.py

```diff
@@ -63,60 +67,74 @@ def run(self) -> None:
         if state != old_state:
             self.freqtrade.notify_status(f'{state.name.lower()}')
 
-            logger.info('Changing state to: %s', state.name)
+            logger.info(f"Changing state to: {state.name}")
             if state == State.RUNNING:
                 self.freqtrade.startup()
 
+            # Reset heartbeat timestamp to log the heartbeat message at
+            # first throttling iteration when the state changes
+            self._heartbeat_msg = 0
+
         if state == State.STOPPED:
             # Ping systemd watchdog before sleeping in the stopped state
             if self._sd_notify:
                 logger.debug("sd_notify: WATCHDOG=1\\nSTATUS=State: STOPPED.")
                 self._sd_notify.notify("WATCHDOG=1\nSTATUS=State: STOPPED.")
 
-            time.sleep(throttle_secs)
+            self._throttle(func=self._process_stopped, throttle_secs=self._throttle_secs)
 
         elif state == State.RUNNING:
             # Ping systemd watchdog before throttling
             if self._sd_notify:
                 logger.debug("sd_notify: WATCHDOG=1\\nSTATUS=State: RUNNING.")
                 self._sd_notify.notify("WATCHDOG=1\nSTATUS=State: RUNNING.")
 
-            self._throttle(func=self._process, min_secs=throttle_secs)
+            self._throttle(func=self._process_running, throttle_secs=self._throttle_secs)
+
+        if self._heartbeat_interval:
+            now = time.time()
+            if (now - self._heartbeat_msg) > self._heartbeat_interval:
+                logger.info(f"Bot heartbeat. PID={getpid()}, "
+                            f"version='{__version__}', state='{state.name}'")
+                self._heartbeat_msg = now
 
         return state
 
-    def _throttle(self, func: Callable[..., Any], min_secs: float, *args, **kwargs) -> Any:
+    def _throttle(self, func: Callable[..., Any], throttle_secs: float, *args, **kwargs) -> Any:
         """
         Throttles the given callable that it
         takes at least `min_secs` to finish execution.
         :param func: Any callable
-        :param min_secs: minimum execution time in seconds
-        :return: Any
+        :param throttle_secs: throttling interation execution time limit in seconds
+        :return: Any (result of execution of func)
         """
-        start = time.time()
+        self.last_throttle_start_time = time.time()
+        logger.debug("========================================")
         result = func(*args, **kwargs)
-        end = time.time()
-        duration = max(min_secs - (end - start), 0.0)
-        logger.debug('Throttling %s for %.2f seconds', func.__name__, duration)
-        time.sleep(duration)
+        time_passed = time.time() - self.last_throttle_start_time
+        sleep_duration = max(throttle_secs - time_passed, 0.0)
+        logger.debug(f"Throttling with '{func.__name__}()': sleep for {sleep_duration:.2f} s, "
+                     f"last iteration took {time_passed:.2f} s.")
```

Comment:
wouldn't we want the "s" be right next to the number?

so "sleep for 2s" - instead of "sleep for 2 s"? or is it intentionally done this way?

Reply:
According to English typographical rules, as Wikipedia manuals say to us (and people at Wikipedia know this pretty well, believe me):

"Use a non-breaking space ({{nbsp}} or &nbsp;) between a number and a unit symbol, or use {{nowrap}", for program code/ASCII output it's blank space. 

Then, "Unit symbols are undotted". 

Btw, 'sec' is also allowed **unit symbol** for 'seconds' **unit name**, I wanted to use it as an alternative here (aligns better with 'min').

"Unit symbols (in any system) are identical in singular and plural." So, no 'secs', 'mins' are allowed, only 'seconds', 'minutes', 's (sec)', 'm (min)'.

See https://en.wikipedia.org/wiki/Wikipedia:Manual_of_Style/Dates_and_numbers#Unit_names_and_symbols.